### DISCUSSION TITLE
CommitShaderResources() frequency

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#231] [Q&A] [answered]
Hey, I just had a few questions about `CommitShaderResources()`. Primarily when and how often it needs called.

In some of the sample files, it mentions it should be called after using `SetPipelineState()`. In my case, I'm using a single global resource signature (`CreatePipelineResourceSignature()`), and I've been calling `CommitShaderResources()` only once per frame, at the beginning of every frame, and have not had any problems to date... until today.

Today, I ran into a situation where I needed to load new textures in during the middle of a frame, and Diligent Engine did not seem to like that I tried to draw polygons after loading new textures without calling `CommitShaderResources()`. Weather that was directly related to the texture loading or some side effect, I'm not sure.

Is there any documentation or general notes about this process that I can read? I'd like to have a firm grasp of the general events that would invalidate resource bindings and require the method to be called. I'm currently exclusively DirectX 12, but would prefer to keep things as compatible as possible for the future, if that matters.

I appreciate any information!

**Comment by TheMostDiligent** [comment, marked as answer]
> In some of the sample files, it mentions it should be called after using SetPipelineState()

No, this is not the case. You can commit resource any time. Note only that binding a PSO that is not compatible with a signature, makes all incompatible SRBs invalid. If you can point to the document, this should be fixed.

> I've been calling CommitShaderResources() only once per frame, at the beginning of every frame, and have not had any problems to date

Yes, this is the exactly correct use case.

> Diligent Engine did not seem to like that

You need call `CommitShaderResources` after you change any variable - the changes are not automatically visible to GPU. If you loaded a new texture and set it to the SRB you committed at the beginning of the frame, this will either have no effect or have some undesired results. Diligent will warn you if you are doing something incorrect though.

**Comment by RobertMtx** [comment]
That's great. I was hoping the behavior would be something along those lines. Thanks for the information.

I'm still not exactly sure which specific thing is triggering the need for the commit. I automatically create shader resource views for my textures as soon as they are created. The textures this happens with are cube maps, and I'm also binding them to a texture array. And these are the first cube maps that are ever loaded, which means I'm probably mapping the texture array for the first time. Maybe even creating a dynamic buffer for the first time (to upload environment map details).

I tried committing right before the binding part, immediately following the texture load, and it still corrects the warning (everything works). But I'm not sure how accurate the exception testing of stale data is. I'm not sure if I can just move my `CommitShaderResources()` function around to see what happens and expect accurate feedback. For example, I'm not sure if Diligent can tell the difference when I call it right-before or right-after a specific thing happens, such as changing SRB variables.

Edit: After a little more testing, I can almost say for certain that the exception was triggered because my newly loaded textures were being binded to my texture array. The reason it still worked when I called commit before the binding section was because there were multiple environment maps going through the whole process, so regardless of where I moved the call, it was being called multiple times before the draw. Doh! So I added a little state when textures are registered to remember to commit before the next draw.

**Reply by TheMostDiligent** [reply]
> I'm still not exactly sure which specific thing is triggering the need for the commit.

Any time you call `Set` or `SetArray` method of the SRB, you have to call `CommitShaderResources()` to make these changes effective and visible to GPU.

Changing the resource content does _not_ require calling the commit method, only when you change the bindings in the SRB.
